### PR TITLE
Clean up ScrollbarPainter

### LINKED ISSUE #107181: ScrollbarPainter clean up
The scrollbar painter has a bit of code rot.

While working on https://github.com/flutter/flutter/issues/13253, I found it hard to follow the painter logic and found several little bugs. This really needs a clean up to make it easier to work with. Some of the bugs I found have to do with padding and margins, these should likely be wrapped in consistent getters to use across the all of the painter methods.

Found bugs:
- #106918 
- #106917 
- #106834 

Fix is in https://github.com/flutter/flutter/pull/107179

### PR DESCRIPTION
**Pull request by Piinks**
Fixes https://github.com/flutter/flutter/issues/107181

This is a cleanup of the ScrollbarPainter. Ended up yak shaving this while working on https://github.com/flutter/flutter/issues/13253. This will make it a bit cleaner, more readable, and easier to work with. 😉 

These need to land first:
- https://github.com/flutter/flutter/pull/107177
- https://github.com/flutter/flutter/pull/107172
- #106835
- #107756

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Review by Piinks — COMMENTED** [review]
In general, this will need to be updated to resolve conflicts from the other changes that were separated from this.

File: packages/flutter/lib/src/widgets/scrollbar.dart [hunk 1]
```diff
@@ -387,23 +387,123 @@ class ScrollbarPainter extends ChangeNotifier implements CustomPainter {
...
+  Rect? _thumbRect;
+  late double _thumbOffset;
+  late double _thumbExtent;
+  void _setThumbExtent() {
+    // Thumb extent reflects fraction of content visible, as long as this
+    // isn't less than the absolute minimum size.
+    // extentInside <= viewportDimension and used for appropriate scaling
+    // accounting for insets & padding.
+    final double fractionVisible = clampDouble(
+      _lastMetrics!.extentInside / _totalContentExtent,
```
[the remark this answers is not in the document]
Comment (Piinks) [reply packages/flutter/lib/src/widgets/scrollbar.dart:404]: I looked into making that change with the padding @xu-baolin and I found it is actually correct. The fractionVisible takes into account the padding because the viewport can extend behind widgets like app bars. There are some `scrollbar_paint_tests` that demonstrate this.
Say there is a cupertino navigation bar like in this test: https://github.com/flutter/flutter/blob/2d771d4966b6a8562bd026807fccbbddecf42e2e/packages/flutter/test/cupertino/scrollbar_paint_test.dart#L64

The `fractionVisible` needs to account for the navigation bar since the `viewportDimension` does not. 

This also reminded me of: https://github.com/flutter/flutter/blob/2d771d4966b6a8562bd026807fccbbddecf42e2e/packages/flutter/lib/src/widgets/scroll_view.dart#L662

That is the padding that was being considered here. I'll add it back to this PR with some comments. It really confused me for a minute, because you are so right. When you read it, it does not look right when computing fractionVisible, but it is. 

All of the other minor bugs have been fixed, so I will try to have this PR updated today and ready to land. It should not require updating any tests since it is meant to just be a clean up. 🤞  😜

**Comment by Hixie** [comment]
test-exempt: refactor with no semantic change

File: packages/flutter/lib/src/widgets/scrollbar.dart [hunk 2]
```diff
@@ -391,23 +391,127 @@ class ScrollbarPainter extends ChangeNotifier implements CustomPainter {
...
+        return textDirection == TextDirection.ltr
+          ? ScrollbarOrientation.right
+          : ScrollbarOrientation.left;
+      }
+      return ScrollbarOrientation.bottom;
+    }
+    return scrollbarOrientation!;
+  }
+
+  void _debugAssertIsValidOrientation(ScrollbarOrientation orientation) {
```
Comment (xu-baolin) [inline packages/flutter/lib/src/widgets/scrollbar.dart:502]: Wrap all the statement body in the assertion scope?

File: packages/flutter/lib/src/widgets/scrollbar.dart [hunk 3]
```diff
@@ -416,7 +520,7 @@ class ScrollbarPainter extends ChangeNotifier implements CustomPainter {
   void update(
     ScrollMetrics metrics,
     AxisDirection axisDirection,
-  ) {
+    ) {
```
Comment (xu-baolin) [inline packages/flutter/lib/src/widgets/scrollbar.dart:523]: nit: Extra spaces?

File: packages/flutter/lib/src/widgets/scrollbar.dart [hunk 4]
```diff
@@ -557,78 +645,41 @@ class ScrollbarPainter extends ChangeNotifier implements CustomPainter {
...
+        || _lastMetrics == null
+        || _lastMetrics!.maxScrollExtent <= _lastMetrics!.minScrollExtent) {
+      return;
+    }
+    // Skip painting if there's not enough space.
+    if (_traversableTrackExtent <= 0) {
+      return;
+    }
+    // Do not paint a scrollbar if the scroll view is infinitely long.
+    // TODO(Piinks): Special handling for infinite scroll views,
```
Comment (xu-baolin) [inline packages/flutter/lib/src/widgets/scrollbar.dart:660]: Looking forward to this being removed one day. ：）

**Review by xu-baolin — APPROVED** [review]
LGTM - with small nits
Nice job, more clear and readable!

File: packages/flutter/lib/src/widgets/scrollbar.dart (hunk 4 above)
Comment (Piinks) [reply packages/flutter/lib/src/widgets/scrollbar.dart:660]: You and me both! I had a proposal for it, I just need to make a lot of adjustments based on the feedback I got in design review. It is on my list to return to once I land the changes for #13253 🎉
